### PR TITLE
PR: Add Panels and PanelsManager, and use it to add LineNumberArea to editor

### PR DESCRIPTION
**Pull request by rlaverde**
Using pyqode code for managing editors panels

Progress
- add pyqode classes
- add panels to codeeditor
- add linenumberarea using panel manager

TODO

- [x] Reconnect all calculation that depends of width linenumberarea (lines commented out)
- [x] Add pyqode copyright and license and links to source files
- [x] Reword, reorder and squash some commits
- [x] Test that everything is working

### CODE REVIEW

**Comment by rlaverde** [comment]
@goanpeca I think this is kind of ready (It will need an squash and reorder of some commits) could you review it?

And what do you think of the deletion of the` LineNumberPanel.update_width` method? It was causing an infinite recursion, and I think that after adding PanelManager that method in not longer needed

**Comment by goanpeca** [comment]
rafa why the tests fail om circleci? do you have the latest master?

**Comment by ccordoba12** [comment]
Tests are failing because CIrcleCI tries to be smart and run without a `circle.yml` file. A simple merge with master will fix that :-)

File: spyder/api/manager.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+        # use the backend manager to start the backend server
+        editor.backend.start(...)
+        editor.backend.send_request(...)
+
+        # use the panels controller to install a panel
+        editor.panels.install(MyPanel(), MyPanel.Position.Right)
+        my_panel = editor.panels.get(MyPanel)
+
+        # and so on
```
Comment (ccordoba12) [inline spyder/api/manager.py]: Please rewrite this docstring to refer to Spyder functionality instead of PyQode one

File: spyder/api/manager.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+    """
+
+    @property
+    def editor(self):
+        """
+        Return a reference to the parent code edit widget.
```
Comment (ccordoba12) [inline spyder/api/manager.py]: We don't code edit widgets, please rewrite

File: spyder/api/manager.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+        """
+        return self._editor()
+
+    def __init__(self, editor):
+        """
+        :param editor: CodeEdit instance to control
```
Comment (ccordoba12) [inline spyder/api/manager.py]: Please rewrite too

File: spyder/api/mode.py [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+        - :meth:`pyqode.core.api.CodeEdit.panels.append`
+
+    Subclasses may/should override the following methods:
+
+        - :meth:`pyqode.core.api.Mode.on_install`
+        - :meth:`pyqode.core.api.Mode.on_uninstall`
+        - :meth:`pyqode.core.api.Mode.on_state_changed`
+
+    ..warning: The mode will be identified by its class name, this means that
+    **there cannot be two modes of the same type on the same editor instance!**
```
Comment (ccordoba12) [inline spyder/api/mode.py]: Please rewrite this docstring too to refer to Spyder objects instead of PyQode ones

File: spyder/api/mode.py [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+    """
+
+    @property
+    def editor(self):
+        """
+        Returns a reference to the parent editor widget.
+
+        **READ ONLY**
+
+        :rtype: pyqode.core.api.code_edit.CodeEdit
```
Comment (ccordoba12) [inline spyder/api/mode.py]: Rewrite

File: spyder/api/mode.py [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+            return None
+
+    @property
+    def enabled(self):
+        """
+        Tells if the mode is enabled,
+        :meth:`pyqode.core.api.Mode.on_state_changed` will be called as soon
+        as the mode state changed.
+
+        :type: bool
```
Comment (ccordoba12) [inline spyder/api/mode.py]: Rewrite

File: spyder/api/mode.py [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+        """
+        Installs the extension on the editor.
+
+        :param editor: editor widget instance
+        :type editor: pyqode.core.api.code_edit.CodeEdit
+
+        .. note:: This method is called by editor when you install a Mode.
+                  You should never call it yourself, even in a subclasss.
+
+        .. warning:: Don't forget to call **super** when subclassing
```
Comment (ccordoba12) [inline spyder/api/mode.py]: Rewrite

File: spyder/api/panel.py [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+    def on_install(self, editor):
+        """
+        Extends :meth:`pyqode.core.api.Mode.on_install` method to set the
+        editor instance as the parent widget.
+
+        .. warning:: Don't forget to call **super** if you override this
+            method!
+
+        :param editor: editor instance
+        :type editor: pyqode.core.api.CodeEdit
```
Comment (ccordoba12) [inline spyder/api/panel.py]: Rewrite

File: spyder/widgets/panels/manager.py [hunk 9]
```diff
@@ -0,0 +1,309 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright © Spyder Project Contributors
+# Copyright © <2013-2016> <Colin Duquesnoy and others, see pyqode/AUTHORS.rst>
+# Licensed under the terms of the MIT License
+# (see spyder/__init__.py for details)
+
+"""
+This module contains the panels controller, responsible of drawing panel
+inside CodeEdit's margins
```
Comment (ccordoba12) [inline spyder/widgets/panels/manager.py]: We don't have CodeEdit, please rewrite

File: spyder/widgets/panels/manager.py [hunk 10]
```diff
@@ -0,0 +1,309 @@
...
+
+def _logger():
+    return logging.getLogger(__name__)
+
+
+class PanelsManager(Manager):
+    """
+    Manages the list of panels and draws them inised the margin of the code
+    edit widget.
+
```
Comment (ccordoba12) [inline spyder/widgets/panels/manager.py]: Remove blank

File: spyder/widgets/panels/manager.py [hunk 11]
```diff
@@ -0,0 +1,309 @@
...
+        panel = self.get(name_or_klass)
+        panel.on_uninstall()
+        panel.hide()
+        panel.setParent(None)
+        return self._panels[panel.position].pop(panel.name, None)
+
+    def clear(self):
+        """
+        Removes all panel from the editor.
+
```
Comment (ccordoba12) [inline spyder/widgets/panels/manager.py]: Remove blank

**Comment by ccordoba12** [comment]
This looks good to me. I'd only like to see better docstrings for files added in this PR, i.e. removing the references to pyqode objects (like `CodeEdit`) and classes, and use the Spyder ones instead :-)

**Comment by rlaverde** [comment]
@ccordoba12 I addressed your comments, but there is a little error, when editor is loaded the panels are drawn over the editor (cropping the first lines), but after scrolling they work fine.

**Comment by rlaverde** [comment]
> but there is a little error, when editor is loaded the panels are drawn over the editor (cropping the first lines)

Fixed, It was missing the override of `CodeEditor.showEvent()` method

File: spyder/api/mode.py [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+    """
+    Base class for editor extensions. An extension is a "thing" that can be
+    installed on an editor to add new behaviours or to modify its appearance.
+
+    At the current states Modes can't be added to an Editor but this class
+    is needed because Panels can be added to an editor.
+
+    A panel (model child class) is added to an editor by using the
+    PanelsManager:
+        - :meth:`pyqode.core.api.CodeEditor.panels.append`
```
Comment (ccordoba12) [inline spyder/api/mode.py]: Please change this line to the corresponding behavior in Spyder

File: spyder/widgets/panels/manager.py [hunk 13]
```diff
@@ -0,0 +1,308 @@
...
+from spyder.api.panel import Panel
+
+
+def _logger():
+    return logging.getLogger(__name__)
+
+
+class PanelsManager(Manager):
+    """
+    Manages the list of panels and draws them inised the margin of the
```
Comment (ccordoba12) [inline spyder/widgets/panels/manager.py]: Misspelling: inised -> inside

**Comment by goanpeca** [comment]
@ccordoba12 ? can we merge this?

File: spyder/api/panel.py [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+    def scrollable(self, value):
+        self._scrollable = value
+
+    def __init__(self, dynamic=False):
+        Mode.__init__(self)
+        QWidget.__init__(self)
+        #: Specifies whether the panel is dynamic. A dynamic panel is a panel
+        #: that will be shown/hidden depending on the context.
+        #: Dynamic panel should not appear in any GUI menu (e.g. no display
+        #: in the panels menu of the notepad example).
```
Comment (ccordoba12) [inline spyder/api/panel.py]: Please remove from this comment the reference to *the notepad example* (which we don't have)

Comment (ccordoba12) [inline spyder/api/panel.py]: Also change `#: ` to `#` in all comments

File: spyder/api/mode.py [hunk 15]
```diff
@@ -0,0 +1,134 @@
...
+
+        .. warning:: Don't forget to call **super** when subclassing
+        """
+        self._editor = editor
+        self.enabled = True
+
+    def on_uninstall(self):
+        """
+        Uninstalls the mode from the editor.
+        """
```
Comment (ccordoba12) [inline spyder/api/mode.py]: Please make this a single line docstring

File: spyder/api/mode.py [hunk 16]
```diff
@@ -0,0 +1,134 @@
...
+
+    @enabled.setter
+    def enabled(self, enabled):
+        if enabled != self._enabled:
+            self._enabled = enabled
+            self.on_state_changed(enabled)
+
+    def __init__(self):
+        #: Mode name/identifier. :class:`spyder.widgets.sourcecode.CodeEditor`
+        # uses that as the attribute key when you install a mode.
```
Comment (ccordoba12) [inline spyder/api/mode.py]: Please make this a docstring

File: spyder/api/mode.py [hunk 17]
```diff
@@ -0,0 +1,134 @@
...
+        """
+        if self._editor is not None:
+            return self._editor
+        else:
+            return None
+
+    @property
+    def enabled(self):
+        """
+        Tells if the mode is enabled,
```
Comment (ccordoba12) [inline spyder/api/mode.py]: Finish this line with a dot and add a blank line to separate it from the description

File: spyder/api/mode.py [hunk 18]
```diff
@@ -0,0 +1,134 @@
...
+import logging
+
+
+def _logger():
+    return logging.getLogger(__name__)
+
+
+class Mode(object):
+    """
+    Base class for editor extensions. An extension is a "thing" that can be
```
Comment (ccordoba12) [inline spyder/api/mode.py]: This should look like

    Base class for editor extensions.

    An extension is a "thing" that can be ...

File: spyder/api/manager.py [hunk 19]
```diff
@@ -0,0 +1,45 @@
...
+    @property
+    def editor(self):
+        """
+        Return a reference to the parent CodeEditor widget.
+        """
+        return self._editor()
+
+    def __init__(self, editor):
+        """
+        :param editor: CodeEditor instance to control
```
Comment (ccordoba12) [inline spyder/api/manager.py]: Make this a single class docstring

File: spyder/api/manager.py [hunk 20]
```diff
@@ -0,0 +1,45 @@
...
+        my_panel = editor.panels.get(MyPanel)
+
+        # and so on
+
+    """
+
+    @property
+    def editor(self):
+        """
+        Return a reference to the parent CodeEditor widget.
```
Comment (ccordoba12) [inline spyder/api/manager.py]: Single line docstring

File: spyder/api/panel.py [hunk 21]
```diff
@@ -0,0 +1,127 @@
...
+    Base class for editor panels.
+
+    A panel is a mode and a QWidget.
+
+    .. note:: Use enabled to disable panel actions and setVisible to change the
+        visibility of the panel.
+    """
+    class Position(object):
+        """
+        Enumerates the possible panel positions
```
Comment (ccordoba12) [inline spyder/api/panel.py]: Single line docstring

File: spyder/api/panel.py [hunk 22]
```diff
@@ -0,0 +1,127 @@
...
+
+        @classmethod
+        def iterable(cls):
+            """ Returns possible positions as an iterable (list) """
+            return [cls.TOP, cls.LEFT, cls.RIGHT, cls.BOTTOM]
+
+    @property
+    def scrollable(self):
+        """
+        A scrollable panel will follow the editor's scroll-bars. Left and right
```
Comment (ccordoba12) [inline spyder/api/panel.py]: Change to

    A scrollable panel will follow the editor's scroll-bars.

    Left and right ...

File: spyder/api/panel.py [hunk 23]
```diff
@@ -0,0 +1,127 @@
...
+        self.setPalette(QApplication.instance().palette())
+        self.setFont(QApplication.instance().font())
+        self.editor.panels.refresh()
+        self._background_brush = QBrush(QColor(
+            self.palette().window().color()))
+        self._foreground_pen = QPen(QColor(
+            self.palette().windowText().color()))
+
+    def paintEvent(self, event):
+        # Fills the panel background using QPalette
```
Comment (ccordoba12) [inline spyder/api/panel.py]: Make this a single line docstring

File: spyder/api/panel.py [hunk 24]
```diff
@@ -0,0 +1,127 @@
...
+            self._background_brush = QBrush(QColor(
+                self.palette().window().color()))
+            self._foreground_pen = QPen(QColor(
+                self.palette().windowText().color()))
+            painter = QPainter(self)
+            painter.fillRect(event.rect(), self._background_brush)
+
+    def setVisible(self, visible):
+        """
+        Shows/Hides the panel
```
Comment (ccordoba12) [inline spyder/api/panel.py]: Dot missing at the end

File: spyder/widgets/panels/manager.py [hunk 25]
```diff
@@ -0,0 +1,308 @@
...
+
+
+def _logger():
+    return logging.getLogger(__name__)
+
+
+class PanelsManager(Manager):
+    """
+    Manages the list of panels and draws them inside the margin of the
+    CodeEditor widget.
```
Comment (ccordoba12) [inline spyder/widgets/panels/manager.py]: Change this to

    Manage the list of panels and draw them inside the margins of
    CodeEditor widgets.

File: spyder/widgets/panels/manager.py [hunk 26]
```diff
@@ -0,0 +1,308 @@
...
+            except KeyError:
+                pass
+            else:
+                return panel
+        raise KeyError(name_or_klass)
+
+    def keys(self):
+        """
+        Returns the list of installed panel names.
+        """
```
Comment (ccordoba12) [inline spyder/widgets/panels/manager.py]: Single line docstring

File: spyder/widgets/panels/manager.py [hunk 27]
```diff
@@ -0,0 +1,308 @@
...
+        return self._modes.keys()
+
+    def values(self):
+        """
+        Returns the list of installed panels.
```
Comment (ccordoba12) [inline spyder/widgets/panels/manager.py]: Single line docstring

File: spyder/widgets/panels/manager.py [hunk 28]
```diff
@@ -0,0 +1,308 @@
...
+        Gets the list of panels attached to the specified zone.
+
+        :param zone: Panel position.
+
+        :return: List of panels instances.
+        """
+        return list(self._panels[zone].values())
+
+    def refresh(self):
+        """ Refreshes the editor panels (resize and update margins) """
```
Comment (ccordoba12) [inline spyder/widgets/panels/manager.py]: Missing dot at the end, remove blanks at the beginning and end of this docstring.

File: spyder/widgets/panels/manager.py [hunk 29]
```diff
@@ -0,0 +1,308 @@
...
+        _logger().log(5, 'refresh_panels')
+        self.resize()
+        self._update(self.editor.contentsRect(), 0,
+                     force_update_margins=True)
+
+    def resize(self):
+        """ Resizes panels """
```
Comment (ccordoba12) [inline spyder/widgets/panels/manager.py]: Removes blanks, missing dot

File: spyder/widgets/panels/manager.py [hunk 30]
```diff
@@ -0,0 +1,308 @@
...
+            size_hint = panel.sizeHint()
+            panel.setGeometry(
+                crect.left(),
+                crect.bottom() - bottom - size_hint.height() - h_offset,
+                crect.width() - w_offset,
+                size_hint.height())
+            bottom += size_hint.height()
+
+    def _update(self, rect, delta_y, force_update_margins=False):
+        """ Updates panels """
```
Comment (ccordoba12) [inline spyder/widgets/panels/manager.py]: Remove blanks, missing dot

File: spyder/widgets/panels/manager.py [hunk 31]
```diff
@@ -0,0 +1,308 @@
...
+                oline, ocol = self._cached_cursor_pos
+                if line != oline or col != ocol or panel.scrollable:
+                    panel.update(0, rect.y(), panel.width(), rect.height())
+                self._cached_cursor_pos = self.editor.get_cursor_line_column()
+        if (rect.contains(self.editor.viewport().rect()) or
+                force_update_margins):
+            self._update_viewport_margins()
+
+    def _update_viewport_margins(self):
+        """ Update viewport margins """
```
Comment (ccordoba12) [inline spyder/widgets/panels/manager.py]: Remove blanks, missing dot

File: spyder/widgets/panels/manager.py [hunk 32]
```diff
@@ -0,0 +1,308 @@
...
+
+        :param position: Margin position. See
+            :class:`spyder.api.Panel.Position`
+        :return: The size of the specified margin
+        :rtype: float
+        """
+        return self._margin_sizes[position]
+
+    def _compute_zones_sizes(self):
+        """ Compute panel zone sizes """
```
Comment (ccordoba12) [inline spyder/widgets/panels/manager.py]: Remove blanks, missing dot

File: spyder/widgets/sourcecode/codeeditor.py [hunk 33]
```diff
@@ -1249,13 +1258,16 @@ def scrollflagarea_paint_event(self, event):
...
         TextEditBaseWidget.resizeEvent(self, event)
         cr = self.contentsRect()
-        self.linenumberarea.setGeometry(\
-                        QRect(cr.left(), cr.top(),
-                              self.linenumberarea.compute_width(), cr.height()))
         self.__set_scrollflagarea_geometry(cr)
+        self.panels.resize()
+
+    def showEvent(self, event):
+        """ Overrides showEvent to update the viewport margins """
```
Comment (ccordoba12) [inline spyder/widgets/sourcecode/codeeditor.py]: Remove blanks, missing dot.

**Comment by ccordoba12** [comment]
I left a lot of comments that ciocheck didn't catch.

**Comment by goanpeca** [comment]
I think it may be a bug on ciocheck I will invetigate further

**Comment by rlaverde** [comment]
> I think it may be a bug on ciocheck I will investigate further

Maybe because they are new files

**Comment by goanpeca** [comment]
>Maybe because they are new files

Yes thats the thing, I dont think I ever used that as a test case, for commited lines (which is the mode we are using now.) Will add an issue

File: spyder/widgets/sourcecode/codeeditor.py (hunk 33 above)
Comment (rlaverde) [inline spyder/widgets/sourcecode/codeeditor.py]: @goanpeca This style issue was in an existing file, and wasn't detected

File: spyder/api/panel.py [hunk 34]
```diff
@@ -0,0 +1,125 @@
...
+import logging
+
+from qtpy.QtWidgets import QWidget, QApplication
+from qtpy.QtGui import QBrush, QColor, QPen, QPainter
+
+from spyder.api.mode import Mode
+
+def _logger():
+    """ Returns module's logger """
+    return logging.getLogger(__name__)
```
Comment (ccordoba12) [inline spyder/api/panel.py]: Please change this (and all other uses) of this `_logger` function to use our own `debug_print` one.

Comment (rlaverde) [inline spyder/api/panel.py]: I haven't notice `_logger` isn't used in the files that I copy, I'm going to delete this functions.

File: spyder/widgets/panels/manager.py [hunk 35]
```diff
@@ -0,0 +1,299 @@
...
+        debug_print('removing panel {}'.format(name_or_klass))
+        panel = self.get(name_or_klass)
+        panel.on_uninstall()
+        panel.hide()
+        panel.setParent(None)
+        return self._panels[panel.position].pop(panel.name, None)
+
+    def clear(self):
+        """
+        Removes all panel from the CodeEditor.
```
Comment (ccordoba12) [inline spyder/widgets/panels/manager.py]: One line docstring.

**Comment by ccordoba12** [comment]
@rlaverde, I left one last comment, the I'd say this is ready ;-)

**Comment by goanpeca** [comment]
Woot :-)

**Pull request merged**